sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/MessageFlux.java

```diff
@@ -0,0 +1,971 @@
+
+                    // there could be pending cancellation signal from the downstream for the operator termination
+                    if (terminateIfCancelled(downstream, null)) {
+                        return;
+                    }
+
+                    // or pending signal from the upstream to terminate the operator with error or completion
+                    // or last drain-loop iteration detected 'non-retriable or retry exhaust' error needing
+                    // operator termination,
+                    if (terminateIfErroredOrUpstreamCompleted(done, downstream, null)) {
+                        return;
+                    }
+
+                    // or pending signal indicating the mediator termination. E.g., receiver detached without
+                    // receiving a message.
+                    if (mediator.queue.isEmpty() && mediator.done) {
+                        mediatorTerminatedAndDrained = true;
+                    }
+                }
+
+                // No need to check 'hasMediator' before accessing 'mediator.isRetryInitiated' since 'mediatorTerminatedAndDrained'
+                // being 'true' means 'mediator' is set.
+                if (mediatorTerminatedAndDrained && !mediator.isRetryInitiated) {
+                    mediator.isRetryInitiated = true;
+                    // The current mediator's queue is drained, and the mediator is terminated, let's close it,
+                    mediator.closeAsync().subscribe();
+                    // and proceed with requesting a new mediator if MessageFlux is in a state to do so.
+                    setTerminationSignalOrScheduleNextMediatorRequest(mediator.error, downstream, mediatorHolder);
+                }
+
+                missed = WIP.addAndGet(this, -missed);
+                if (missed == 0) {
+                    break;
+                }
+                // The next serialized drain-loop iteration to process missed signals arrived during last iteration.
+            }
+            // End: serialized drain-loop.
+        }
+
+        /**
+         * CONTRACT: Never invoke from the outside of serialized drain-loop.
+         * <br/>
+         * See if downstream signaled cancellation to terminate the operator, if so, react to the cancellation.
+         *
+         * @param downstream the downstream.
+         * @param messageDropped the message that gets dropped if cancellation was signaled.
+         * @return true if canceled, false otherwise.
+         */
+        private boolean terminateIfCancelled(CoreSubscriber<? super Message> downstream, Message messageDropped) {
+            if (cancelled) {
+                Operators.onDiscard(messageDropped, downstream.currentContext());
+                upstream.cancel();
+                mediatorHolder.freeze();
+                return true;
+            }
+            return false;
+        }
+
+        /**
+         * CONTRACT: Never invoke from the outside of serialized drain-loop.
```

Comment:
is there a way to check here and `assert` or throw? (and similar question in other places)

Reply:
Such a check is hard to do. All these _private_ methods are extensions of the drain loop, so the best we can do is to make notes for future developers. We hope someone later want to change this operator understands the standard established for general Reactor-RxJava operator implementation. I've linked the technical series on operator implementation as the Java-Doc for drain-loop.

Reply:
All these are _private_ methods to support the drain-loop method in the same class, i.e., invoked only from the drain-loop method. If someone accidentally edits and breaks this contract, then catching that runtime is hard. The best we can do is to add detailed design notes for future developers and necessary references to operator implementation practices (which we added).